presto-main/src/main/java/com/facebook/presto/operator/scalar/JoniRegexpReplaceLambdaFunction.java

```diff
@@ -0,0 +1,112 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator.scalar;
+
+import com.facebook.presto.spi.PageBuilder;
+import com.facebook.presto.spi.block.Block;
+import com.facebook.presto.spi.block.BlockBuilder;
+import com.facebook.presto.spi.function.Description;
+import com.facebook.presto.spi.function.LiteralParameters;
+import com.facebook.presto.spi.function.ScalarFunction;
+import com.facebook.presto.spi.function.SqlNullable;
+import com.facebook.presto.spi.function.SqlType;
+import com.facebook.presto.sql.gen.lambda.UnaryFunctionInterface;
+import com.facebook.presto.type.JoniRegexpType;
+import com.google.common.collect.ImmutableList;
+import io.airlift.joni.Matcher;
+import io.airlift.joni.Option;
+import io.airlift.joni.Regex;
+import io.airlift.joni.Region;
+import io.airlift.slice.DynamicSliceOutput;
+import io.airlift.slice.Slice;
+import io.airlift.slice.SliceOutput;
+
+import static com.facebook.presto.spi.type.VarcharType.VARCHAR;
+
+@ScalarFunction("regexp_replace")
+@Description("replaces substrings matching a regular expression using a lambda function")
+public final class JoniRegexpReplaceLambdaFunction
+{
+    private final PageBuilder pageBuilder = new PageBuilder(ImmutableList.of(VARCHAR));
+
+    @LiteralParameters("x")
+    @SqlType("varchar")
+    @SqlNullable
+    public Slice regexpReplace(
+            @SqlType("varchar") Slice source,
+            @SqlType(JoniRegexpType.NAME) Regex pattern,
+            @SqlType("function(array(varchar), varchar(x))") UnaryFunctionInterface replaceFunction)
+    {
+        Matcher matcher = pattern.matcher(source.getBytes());
+        if (matcher.search(0, source.length(), Option.DEFAULT) == -1) {
+            return source;
+        }
+
+        SliceOutput output = new DynamicSliceOutput(source.length());
+
+        // Prepare a BlockBuilder that will be used to create the target block
+        // that will be passed to the lambda function.
+        if (pageBuilder.isFull()) {
+            pageBuilder.reset();
+        }
+        BlockBuilder blockBuilder = pageBuilder.getBlockBuilder(0);
+
+        int groupCount = pattern.numberOfCaptures();
+        int appendPosition = 0;
+        int nextStart = 0;
+
+        do {
+            // nextStart is the same as the last appendPosition, unless the last
```

Comment:
wrap 'unless the last' to the next line with 'match was zero-width.' Then wrap after period if necessary